zkevm-circuits/src/evm_circuit/execution/blockhash.rs

```diff
@@ -0,0 +1,207 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{N_BYTES_U64, N_BYTES_WORD},
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            from_bytes,
+            math_gadget::{IsEqualGadget, LtGadget},
+            CachedRegion, Cell, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::BlockContextFieldTag,
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{Field, ToLittleEndian, ToScalar};
+use gadgets::util::not;
+use halo2_proofs::plonk::Error;
+
+#[derive(Clone, Debug)]
+pub(crate) struct BlockHashGadget<F> {
+    same_context: SameContextGadget<F>,
+    block_number: RandomLinearCombination<F, N_BYTES_U64>,
+    current_block_number: Cell<F>,
+    block_hash: RandomLinearCombination<F, N_BYTES_WORD>,
+    block_lt: LtGadget<F, N_BYTES_U64>,
+    diff_lt: LtGadget<F, 2>,
+    is_valid: IsEqualGadget<F>,
```

Comment:
You don't need this as you can construct the expression only using `block_lt` and `diff_lt`